common: support s390 and unknown architectures in spin-wait loop

smithfarm opened the pull request:
Fixes: http://tracker.ceph.com/issues/19492
Signed-off-by: Nathan Cutler <ncutler@suse.com>

smithfarm commented:
@badone @tchaikov The jury is still out on whether s390 has such an assembler instruction. This is one way the issue could be dealt with - are we sure we want to break builds for architectures not recognized in this `#if` block, though? Would it make sense to put the old code into the `#else` ?

tchaikov commented:
> are we sure we want to break builds for architectures not recognized in this #if block, though? 

i am not sure now. maybe we should `#else` them.

smithfarm commented:
@kefu So, like this?

tchaikov commented on src/common/simple_spin.cc:
```diff
@@ -18,6 +18,11 @@
 #include <stdint.h>
 #include <pthread.h>
 
+#if !defined(__i386__) && !defined(__x86_64__) && !defined(__arm__) && !defined(__aarch64__) && !defined(__powerpc__) && !defined(__ppc__) 
```
i feel guilty at seeing this. but it works!

smithfarm commented:
One colleague suggested the detail namespace idiom, so I added a commit for that. Another colleague wrote:

> I don't know s390, but looks like in kernel we have just a simple compiler barrier for s390 implementation of cpu_relax() (which is done by "asm pause" on x86).
> 
> The compiler barrier is:
> 
> #define barrier() __asm__ __volatile__("": : :"memory")

So I added a commit for that as well. Please re-review.

smithfarm commented:
Pushed `wip-19492` to ceph-ci.

smithfarm commented:
Builds fine on shaman, also builds successfully on all four architectures (x86_64, aarch64, ppc64le, and s390x) in OBS.

badone commented:
I'd be interested to see the assembly produced for that on s390(x) (even though I know little about the arch). On x86_64 at least it appears to produce no instructions, https://godbolt.org/g/i6vrMP If that is what we are using in the kernel (I looked originally and this is what I found also) I guess it's the best we can do. I puzzle over what sort of DAS s390(x) ceph installations are using, but I guess that's a discussion for another forum?

tchaikov commented on src/common/simple_spin.cc:
```diff
@@ -18,25 +18,42 @@
+
+void simple_spin_lock(simple_spinlock_t *lock)
+{
+  while(1) {
```
could take this chance to add a space after `while`.

smithfarm replied:
done

tchaikov commented on src/common/simple_spin.cc:
```diff
@@ -18,25 +18,42 @@
 #include <pthread.h>
 
-void simple_spin_lock(simple_spinlock_t *lock)
+namespace detail {
```
i don't think `detail` namespace would help here, but an anonymous namespace does. and would be better if we can rename `delay` to something like `cpu_relax` or just `pause`.

smithfarm replied:
ok, done

smithfarm commented:
@badone The purpose of the s390 build is to enable mainframe clients to access external Ceph clusters over the network.

smithfarm commented:
@badone Regarding `"": : :"memory"` I found this: https://en.wikipedia.org/wiki/Memory_ordering#Compile-time_memory_barrier_implementation

smithfarm commented:
As agreed on IRC, pushed latest version to ceph-ci and will run it through `rados/thrash` for validation.

smithfarm commented:
`teuthology-suite -k distro --priority 101 --suite rados/thrash --subset $(expr $RANDOM % 2000)/2000 --email ncutler@suse.com --ceph wip-19492 --machine-type smithi`

**waiting for build**
